packages/zoe/src/contracts/autoswap.js

```diff
@@ -23,245 +22,233 @@ import '../../exported';
-        );
+  const removeLiquidityExpected = harden({
+    want: { TokenA: null, TokenB: null },
+    give: { Liquidity: null },
+  });
 
-      const makeSwapInvite = () =>
-        zcf.makeInvitation(checkHook(swapHook, swapExpected), 'autoswap swap');
+  const swapExpected = {
+    want: { Out: null },
+    give: { In: null },
+  };
 
-      /**
-       * `getCurrentPrice` calculates the result of a trade, given a certain amount
-       * of digital assets in.
-       * @param {Amount} amountIn - the amount of digital
-       * assets to be sent in
-       */
-      const getCurrentPrice = (amountIn, brandOut) => {
-        const inputReserve = getPoolAmount(amountIn.brand).value;
-        const outputReserve = getPoolAmount(brandOut).value;
-        const outputValue = getInputPrice(
-          harden({
-            inputValue: amountIn.value,
-            inputReserve,
-            outputReserve,
-          }),
-        );
-        return zcf.getAmountMath(brandOut).make(outputValue);
-      };
+  const makeAddLiquidityInvite = () =>
+    zcf.makeInvitation(
+      assertProposalKeywords(addLiquidityHandler, addLiquidityExpected),
+      'autoswap add liquidity',
+    );
 
-      const getPoolAllocation = () =>
-        zcf.getCurrentAllocation(poolHandle, brandKeywordRecord);
+  const makeRemoveLiquidityInvite = () =>
+    zcf.makeInvitation(
+      assertProposalKeywords(removeLiquidityHandler, removeLiquidityExpected),
+      'autoswap remove liquidity',
+    );
 
-      zcf.initPublicAPI(
-        harden({
-          getCurrentPrice,
-          getLiquidityIssuer: () => liquidityIssuer,
-          getPoolAllocation,
-          makeSwapInvite,
-          makeAddLiquidityInvite,
-          makeRemoveLiquidityInvite,
-        }),
-      );
+  const makeSwapInvite = () =>
+    zcf.makeInvitation(
+      assertProposalKeywords(swapHandler, swapExpected),
+      'autoswap swap',
+    );
 
-      return makeAddLiquidityInvite();
-    });
+  /**
+   * `getCurrentPrice` calculates the result of a trade, given a certain amount
+   * of digital assets in.
+   * @param {Amount} amountIn - the amount of digital
+   * assets to be sent in
+   * @param brandOut - The brand of asset desired
+   */
+  const getCurrentPrice = (amountIn, brandOut) => {
+    const inputReserve = getPoolAmount(amountIn.brand).value;
+    const outputReserve = getPoolAmount(brandOut).value;
+    const outputValue = getInputPrice(
+      harden({
+        inputValue: amountIn.value,
+        inputReserve,
+        outputReserve,
+      }),
+    );
+    return zcf.getAmountMath(brandOut).make(outputValue);
+  };
+
+  const getPoolAllocation = poolSeat.getCurrentAllocation;
+
+  const publicFacet = harden({
+    getCurrentPrice,
+    getLiquidityIssuer: () => liquidityIssuer,
+    getPoolAllocation,
+    makeSwapInvite,
+    makeAddLiquidityInvite,
+    makeRemoveLiquidityInvite,
   });
+
+  return { publicFacet, creatorFacet: publicFacet };
```

Comment:
Ooo nice, I like the concept of passing the publicFacet back as the creatorFacet. We might consider always passing the publicFacet back to the creator too.